src/Parse.idr

```diff
@@ -38,56 +41,101 @@ fromVMax {m} vm = go lteRefl vm
-          let (mx**vx) = toVMax (x :: toVect nel) in
-          (mx ** con $ map (\(_**(lte,td)) => weakenTDef td mx lte)
-                           (fromVMax vx))
-        ) $
-        parens (rand (withSpaces (char sym))
-           (map2 {a=Parser' _} {b=Parser' _}
-                 (\p, q => and p q)
-                 (map {a=Parser' _} withSpaces rec)
-                 (map {a=Parser' _} (nelist . withSpaces) rec)))
+tdef =
+   fix (Parser' (n ** TDef n)) $ \rec =>
+   withSpaces $
+   alts [ guardM (\(mp, nam) => lookup nam mp) $ mand (lift get) alphas
+        , cmap (Z ** T0) $ string "0"
+        , cmap (Z ** T1) $ string "1"
+        , nary rec '*' TProd
+        , nary rec '+' TSum
+        , map (\n => (S n ** TVar $ last {n})) $
+            parens (rand (withSpaces (string "var")) (withSpaces decimalNat))
+        , randbindm
+            (guardM {b=(String, (n ** TDef n))}
+                    (\(nam, nel) =>
+                     let vs : Vect (length nel) (n : Nat ** (String, TDef n)) =
+                           -- push names under sigma to fit in VMax
+                           map (\(nm,(n**td)) => (n**(nm,td))) $ toVect nel
+                         (mx**vx) = toVMax vs
+                       in
+                     case mx of
+                       Z => Nothing
+                       S m => Just (nam, (m ** TMu nam $ map (\(_**(lte,nm,td)) => (nm, weakenTDef td (S m) lte))
+                                                             (fromVMax vx)))
+                    ) $
+             parens (rand (withSpaces (string "mu"))
+                          (and (withSpaces alphas)
+                               (map {a=Parser' _} (\t => nelist $ withSpaces $ parens $ and (withSpaces alphas) t)
+                                                  rec))))
+            (\(nam, mu) => (lift $ modify $ insert nam mu) *> pure mu)
+        , randbindm
+            (parens (rand (withSpaces (string "name")) (and (withSpaces alphas) (map {a=Parser' _} withSpaces rec))))
+            (\(nm, (n**td)) => (lift $ modify $ insert nm (n**td)) *> pure (n ** TName nm td))
+        ]
+ where
+ nary : All (Box (Parser' (n ** TDef n))
+         :-> Cst  Char
+         :-> Cst ({k, m : Nat} -> Vect (2+k) (TDef m) -> TDef m)
+         :->      Parser' (n ** TDef n))
+ nary rec sym con =
+   map (\(x,nel) =>
+         -- find the upper bound and weaken all elements to it
+         let (mx**vx) = toVMax (x :: toVect nel) in
+         (mx ** con $ map (\(_**(lte,td)) => weakenTDef td mx lte)
+                          (fromVMax vx))
+       ) $
+       parens (rand (withSpaces (char sym))
+          (map2 {a=Parser' _} {b=Parser' _}
+                (\p, q => and p q)
+                (map {a=Parser' _} withSpaces rec)
+                (map {a=Parser' _} (nelist . withSpaces) rec)))
+
+tdefRec : All (Parser' (n ** TDef n))
+tdefRec = fix _ $ \rec => map (\(a, ma) => fromMaybe a ma) $ andopt tdef rec
```

Comment:
If I read this correctly, this will parse a sequence of `TDef`s, and return the last one (that parsed)? Perhaps there would be some value in also defining the function that returns a list of `Tdef`s?

Reply:
Would it make sense to add @fredrikNordvallForsberg's remark as a doc comment?